execution_engine/src/storage/global_state/scratch.rs

```diff
@@ -0,0 +1,495 @@
+        let ret = match read::<Key, StoredValue, lmdb::RoTransaction, LmdbTrieStore, Self::Error>(
+            correlation_id,
+            &txn,
+            self.trie_store.deref(),
+            &self.root_hash,
+            key,
+        )? {
+            ReadResult::Found(value) => {
+                self.cache.write().unwrap().insert(*key, value.clone());
+                Some(value)
+            }
+            ReadResult::NotFound => None,
+            ReadResult::RootNotFound => panic!("ScratchGlobalState has invalid root"),
+        };
+        txn.commit()?;
+        Ok(ret)
+    }
+
+    fn read_with_proof(
+        &self,
+        correlation_id: CorrelationId,
+        key: &Key,
+    ) -> Result<Option<TrieMerkleProof<Key, StoredValue>>, Self::Error> {
+        let txn = self.environment.create_read_txn()?;
+        let ret = match read_with_proof::<
+            Key,
+            StoredValue,
+            lmdb::RoTransaction,
+            LmdbTrieStore,
+            Self::Error,
+        >(
+            correlation_id,
+            &txn,
+            self.trie_store.deref(),
+            &self.root_hash,
+            key,
+        )? {
+            ReadResult::Found(value) => Some(value),
+            ReadResult::NotFound => None,
+            ReadResult::RootNotFound => panic!("LmdbWithCacheGlobalState has invalid root"),
+        };
+        txn.commit()?;
+        Ok(ret)
+    }
+
+    fn keys_with_prefix(
+        &self,
+        correlation_id: CorrelationId,
+        prefix: &[u8],
+    ) -> Result<Vec<Key>, Self::Error> {
+        let txn = self.environment.create_read_txn()?;
+        let keys_iter = keys_with_prefix::<Key, StoredValue, _, _>(
+            correlation_id,
+            &txn,
+            self.trie_store.deref(),
+            &self.root_hash,
+            prefix,
+        );
+        let mut ret = Vec::new();
+        for result in keys_iter {
+            match result {
+                Ok(key) => ret.push(key),
+                Err(error) => return Err(error),
+            }
+        }
+        txn.commit()?;
+        Ok(ret)
+    }
+}
+
+impl CommitProvider for ScratchGlobalState {
+    /// State hash returned is the one provided, as we do not write to lmdb with this kind of global
+    /// state.
+    fn commit(
+        &self,
+        _correlation_id: CorrelationId,
+        state_hash: Digest,
+        effects: AdditiveMap<Key, Transform>,
+    ) -> Result<Digest, Self::Error> {
+        for (key, transform) in effects.into_iter() {
+            let cached_value = self.cache.read().unwrap().get(&key).cloned();
+            let value = match (cached_value, transform) {
+                (None, Transform::Write(new_value)) => new_value,
+                (None, transform) => {
+                    error!(
+                        ?key,
+                        ?transform,
+                        "Key not found while attempting to apply transform"
+                    );
```

Comment:
Ideally, in this case, we would read from the underlying LMDB and apply the transform on top if it.

Reply:
This code was lifted from the top-level `commit` method where effects are collapsed. I believe this is correct because we would have had to read the value into cache in order to write any changes to it. So somehow if we have a transform that doesn't correspond to a stored value that we already have cached, we're doing something wrong and this should be an error.